libs/cache/file_cache.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	log.Debugf(ctx, "[Local Cache] using cache key: %s\n", cacheKey)
+	fc.addTelemetryMetric("local.cache.attempt")
+
+	cachePath := fc.getCachePath(cacheKey)
+
+	// Try to read from disk cache
+	cachedData, cacheExists := fc.readFromCache(ctx, cachePath)
+
+	// Record metrics
+	if cacheExists {
+		log.Debugf(ctx, "[Local Cache] cache hit\n")
+		fc.addTelemetryMetric("local.cache.hit")
+
+		// If cache is enabled, return the cached value
+		if fc.cacheEnabled {
+			return cachedData, nil
+		}
+	} else {
+		log.Debugf(ctx, "[Local Cache] cache miss, computing\n")
+		fc.addTelemetryMetric("local.cache.miss")
+	}
+
+	// Acquire lock to prevent concurrent computations and writes for the same cache key
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+
+	// Compute the value and measure timing
+	start := time.Now()
+	result, err := compute(ctx)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] error while computing: %v\n", err)
+		fc.addTelemetryMetric("local.cache.error")
+		return result, err
+	}
+
+	// Record duration metrics
+	if fc.metrics != nil {
+		computeDuration := time.Since(start)
+		fc.metrics.SetDurationValue("local.cache.compute_duration", computeDuration)
+	}
+
+	log.Debugf(ctx, "[Local Cache] computed and stored result\n")
+
+	// Write to disk cache (failures are silent - cache write errors don't affect the result)
+	fc.writeToCache(ctx, cachePath, result)
+
+	return result, nil
+}
+
+// readFromCache attempts to read and deserialize data from the cache file.
+// Expiry is checked using file modification time for consistency with cleanup.
+func (fc *FileCache[T]) readFromCache(ctx context.Context, cachePath string) (T, bool) {
+	var zero T
+
+	// Check file modification time for expiry
+	info, err := os.Stat(cachePath)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] failed to stat cache file: %v\n", err)
+		return zero, false
+	}
+
+	age := time.Since(info.ModTime())
+	expiryDuration := time.Duration(fc.expiryMinutes) * time.Minute
+	if age > expiryDuration {
+		return zero, false
+	}
+
+	// Read and deserialize the data
+	data, err := os.ReadFile(cachePath)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] failed to read cache file: %v\n", err)
+		return zero, false
+	}
+
+	var result T
+	if err := json.Unmarshal(data, &result); err != nil {
+		log.Debugf(ctx, "[Local Cache] failed to deserialize data: %v\n", err)
+		return zero, false
+	}
+
+	return result, true
+}
+
+// writeToCache serializes and writes data to the cache file.
+// Expiry is tracked by file modification time, not stored in the file.
+func (fc *FileCache[T]) writeToCache(ctx context.Context, cachePath string, data any) {
+	// Serialize the data directly
+	serializedData, err := json.Marshal(data)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] failed to serialize data: %v\n", err)
+		return // Silently fail on serialization errors
+	}
+
+	// Write to cache file - the mtime will be used to track expiry
+	err = os.WriteFile(cachePath, serializedData, 0o600)
```

Comment:
This is tricky; the write must be atomic or we risk writing a partial file and returning an invalid cache entry.

You can write to a temporary file in the same directory (with https://pkg.go.dev/os#CreateTemp) and rename after the write has completes. The rename is guaranteed to be atomic. It can still fail if multiple processes try to create the same time so failures must also be ignored/logged.